src/test/run-pass/zero-sized-string-push.rs

```diff
@@ -0,0 +1,32 @@
+// Copyright 2013 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+use std::string::String;
+
+fn main() {
+    const N: usize = 8;
+
+    // Test that for all possible sequences of push,
+    // we end up with a string of the correct size
+
+    for len in 0..N {
+        let mut tester = String::with_capacity(len);
+        assert_eq!(tester.len(), 0);
+        assert!(tester.capacity() >= len);
+        for case in 0..(1 << len) {
+            assert_eq!(tester.len(), 0);
+            for bit in 0..len {
+                tester.push_str("");
+            }
+            assert_eq!(tester.len(), len);
+            tester.clear();
+        }
+    }
+}
```

Comment:
Regardless the test is incorrect, the length of the string should never increase.